distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala

```diff
@@ -962,6 +967,61 @@ object Lifecycle extends LifecycleCatsInstances {
     }
   }
 
+  @inline
+  private final def foldMImpl[F[_], A, B](
+    self: Lifecycle[F, A]
+  )(failure: Throwable => Lifecycle[F, B],
+    success: A => Lifecycle[F, B],
+  )(implicit F: QuasiIO[F]
+  ): Lifecycle[F, B] = {
+    import QuasiIO.syntax._
+    new Lifecycle[F, B] {
+      override type InnerResource = AtomicReference[List[() => F[Unit]]]
+
+      private[this] def bracketAppendFinalizer[a, b](finalizers: InnerResource)(lifecycle: Lifecycle[F, a])(use: lifecycle.InnerResource => F[b]): F[b] = {
+        F.bracket(
+          acquire = lifecycle.acquire.flatMap {
+            a =>
+              F.maybeSuspend {
+                // can't use `.updateAndGet` because of Scala.js
+                var oldValue = finalizers.get()
+                while (!finalizers.compareAndSet(oldValue, (() => lifecycle.release(a)) :: oldValue)) {
+                  oldValue = finalizers.get()
+                }
+                a
+              }
+          }
+        )(release = _ => F.unit)(
+          use = use
+        )
+      }
+
+      override def acquire: F[InnerResource] = F.maybeSuspend(new AtomicReference(Nil))
+      override def release(finalizers: InnerResource): F[Unit] = F.suspendF(F.traverse_(finalizers.get())(_.apply()))
+      override def extract[C >: B](finalizers: InnerResource): Either[F[C], C] = Left {
+        bracketAppendFinalizer(finalizers)(self) {
+          inner1: self.InnerResource =>
+            F.definitelyRecover {
```

Comment:
This will not run the failure handler if `self.acquire` failed. Current `Lifecycle` encoding is roughly similar to [old pre-RC19 ZManaged encoding](https://github.com/zio/zio/blob/v1.0.0-RC18/core/shared/src/main/scala/zio/ZManaged.scala) (where `Lifecycle#extract` = `Reservation#acquire`, `Lifecycle#acquire` = `ZManaged#reservation`), you may be able to look at their implementation of [foldM](https://github.com/zio/zio/blob/v1.0.0-RC18/core/shared/src/main/scala/zio/ZManaged.scala#L396) for tips.

Reply:
Yeah, I've taken inspiration there, but seems I misunderstood some parts. Thanks for the review and you notes about implementation!

Reply:
Done.